### ISSUE TITLE
run index not carried over from source data to derivatives

### ISSUE THREAD

**Issue by satra** [nipreps/fmriprep#2666]
### What happened?

filing separate issue instead of contaminating the single run thread #2365

the source data had `run-01`, the outputs are written with `run-1`

>source: /om2/scratch/Sat/dclb/mct/sub-MM240/ses-baseline/func/sub-MM240_ses-baseline_task-nback_rec-moco_run-01_bold.nii.gz
derivative: /om2/scratch/Sat/dclb/mct/derivatives/fmriprep/sub-MM240/ses-baseline/func/sub-MM240_ses-baseline_task-nback_rec-moco_run-1_space-MNI152NLin6Asym_res-2_desc-preproc_bold.nii.gz

and this results in filtering for events to do nilearn-based analysis.

### What command did you use?

```shell
cmd="singularity run -e -B /om2 -B ${scratch}:/workdir -B ${bids_dir}:/input:ro -B ${outdir}:/out ${SING_IMG} /input /out participant --participant_label ${subject} -w /workdir --cifti-output 91k --mem_mb 40000 --output-spaces MNI152NLin6Asym:res-2 anat --fs-subjects-dir /out/freesurfer-${session}/ --bids-filter-file bids-filter-${session}.json --bids-database-dir ${myscratch}/mct/derivatives/bids_db/ --nprocs 16 --omp-nthreads 8"
```

### What version of fMRIPrep are you running?

20.2.6

### How are you running fMRIPrep?

Singularity

### Is your data BIDS valid?

Yes

### Are you reusing any previously computed results?

No

### Additional information / screenshots

cc: @burdinskid13

**Comment by effigies** [comment]
The issue here is that PyBIDS interprets run indices as ints (I believe because people would query `run=1` instead of `run="01"`). I don't see a good way to change this without breaking things significantly for somebody. We also don't want to monkeypatch or otherwise work around PyBIDS in a way that will require constant maintenance, so I've not considered this very plausible.

But here's a thought: We could add a special entity to the BIDSLayout config called something like `run_pad` with regex `run-\(0*\)\d+_` and make it an optional part of the path patterns.

In any case, I'm not sure we could consider this a bug fix; it's been part of fMRIPrep for years now, and would result in a change of derivatives that are not wrong. I am re-classifying this as a feature targetting 22.0, though happy to hear arguments.

xref https://github.com/nipreps/fmriprep/issues/2655

**Comment by satra** [comment]
i understand but you do realize how this breaks downstream tools like nilearn and fitlins when we are dealing with arbitrary datasets, especially since the producer of the rawdataset or the derivative has technically violated anything. 

re: pybids: it is also perhaps about idempotent inverse transforms. can pybids read and write a layout that doesn't change the names. at present it cannot because of things like this.  so i'm not sure it's about monkeypatching if one thinks through the lens of a serialization format, which in an ideal world should be able to maintain an identity transform. unfortunately, it's not roundtrippable as the name generated would look different.

> it's been part of fMRIPrep for years now, and would result in a change of derivatives that are not wrong.

obviously this is an edge case, but nonetheless one that exists. perhaps it says more about what people are doing post fmriprep rather than the result of fmriprep. one could perhaps look through openneuro and see how many datasets this would affect.

**Comment by effigies** [comment]
Once bids-standard/pybids#819 is merged, we should be able to resolve this in 22.0.
